# Проект по SQL

# Анализ базы данных книг

Ваша компания решила быть купить крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг.

# Задачи:

- Посчитать, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитать количество обзоров и среднюю оценку;
- Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — с целью исключить из анализа брошюры;
- Определить автора с самой высокой средней оценкой книг, необходимо учесть только книги с 50 и более оценками;
- Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

# План работы:
- 1 Подготовка рабочей среды и загрузка данных
    - 1.1 Импорт библиотек
    - 1.2 Установка параметров
    - 1.3 Проверка содержимого
- 2 Выполнение задач исследования
    - 2.1 Посчитать, сколько книг вышло после 1 января 2000 года 
    - 2.2 Для каждой книги посчитать количество обзоров и среднюю оценку
    - 2.3 Определить издательство, которое выпустило наибольшее число книг толще 50 страниц — с целью исключить из анализа брошюры
    - 2.4 Определить автора с самой высокой средней оценкой книг, необходимо учесть только книги с 50 и более оценками
    - 2.5 Посчитайть среднее количество обзоров от пользователей, которые поставили больше 48 оценок
- 3 Вывод

## Подготовка рабочей среды и загрузка данных

### Импорт библиотек

In [29]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

### Установка параметров

Параметры подключения удалены согласно соглашению об использованию данных с Яндекс Практикум.

### Проверка содержимого

#### Таблица books

In [31]:
query = '''SELECT * FROM books LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [32]:
query = '''
SELECT COUNT(*) 
FROM books 
'''
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'В базе {result.values[0][0]} книг.')

В базе 1000 книг.


In [33]:
query = '''
SELECT COUNT(*) 
FROM books 
WHERE book_id IS NULL 
OR author_id IS NULL 
OR title IS NULL 
OR publication_date IS NULL
OR publisher_id IS NULL
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0      0

Пропусков в данных не обнаружено

In [34]:
query = '''
SELECT COUNT(*) AS counted 
FROM books 
GROUP BY author_id, title
ORDER BY counted DESC
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

counted
0        2
1        1
2        1
3        1
4        1

В данных таблицы всего один дубликат по полям автор и название, при этом при добавлении издания этот дубликат исчезает.

#### Таблица authors

In [35]:
query = '''SELECT * FROM authors LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [36]:
query = '''
SELECT COUNT(*) 
FROM authors 
'''
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'В базе {result.values[0][0]} авторов.')

В базе 636 авторов.


In [37]:
query = '''
SELECT COUNT(*) 
FROM authors 
WHERE author IS NULL 
OR author_id IS NULL 

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0      0

Пропусков в данных не обнаружено

In [38]:
query = '''
SELECT COUNT(*) AS counted 
FROM authors 
GROUP BY author_id, author
ORDER BY counted DESC
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

counted
0        1
1        1
2        1
3        1
4        1

Дубликатов в таблице authors не обнаружено

#### Таблица ratings

In [39]:
query = '''SELECT * FROM ratings LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [40]:
query = '''
SELECT COUNT(*) 
FROM ratings 
'''
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'В базе {result.values[0][0]} оценок.')

В базе 6456 оценок.


In [41]:
query = '''
SELECT COUNT(*) 
FROM ratings 
WHERE rating_id IS NULL 
OR book_id IS NULL 
OR username IS NULL 
OR rating IS NULL 
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0      0

Пропусков в данных не обнаружено

In [42]:
query = '''
SELECT COUNT(*) AS counted 
FROM ratings 
GROUP BY username, rating,book_id
ORDER BY counted DESC
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

counted
0        1
1        1
2        1
3        1
4        1

Дубликатов в данных таблицы не обнаружено

#### Таблица reviews

In [43]:
query = '''SELECT * FROM reviews LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

In [44]:
query = '''
SELECT COUNT(*) 
FROM reviews 
'''
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'В базе {result.values[0][0]} обзоров.')

В базе 2793 обзоров.


In [45]:
query = '''
SELECT COUNT(*) 
FROM reviews 
WHERE review_id IS NULL 
OR book_id IS NULL 
OR username IS NULL 
OR text IS NULL 
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0      0

Пропусков в данных не обнаружено

In [46]:
query = '''
SELECT COUNT(*) AS counted 
FROM reviews 
GROUP BY username, text
ORDER BY counted DESC
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

counted
0        1
1        1
2        1
3        1
4        1

Дубликатов в данных таблицы не обнаружено

#### Таблица publishers

In [47]:
query = '''SELECT * FROM publishers LIMIT 5'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [48]:
query = '''
SELECT COUNT(*) 
FROM publishers 
'''
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'В базе {result.values[0][0]} изданий.')

В базе 340 изданий.


In [49]:
query = '''
SELECT COUNT(*) 
FROM publishers 
WHERE publisher_id IS NULL 
OR publisher IS NULL 

'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

count
0      0

Пропусков в данных не обнаружено

In [50]:
query = '''
SELECT COUNT(*) AS counted 
FROM publishers 
GROUP BY publisher_id, publisher
ORDER BY counted DESC
LIMIT 5
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

counted
0        1
1        1
2        1
3        1
4        1

Дубликатов в данных таблицы не обнаружено

### Вывод по разделу

Дубликатов и пропусков в таблице не обнаружено.

В таблицах следующее количество записей:
- books: 1000
- authors: 636
- ratings: 6456 
- reviews: 2793 
- publishers: 340 

## Выполнение задач исследования

### Посчитайте, сколько книг вышло после 1 января 2000 года;

In [51]:
query = '''SELECT count(book_id)
FROM books
WHERE (publication_date::date > '2000-01-1')'''
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'Число книг в базе данных, вышедших после 01.01.2000: {result.values[0][0]} единиц.')

Число книг в базе данных, вышедших после 01.01.2000: 819 единиц.


### Для каждой книги посчитайте количество обзоров и среднюю оценку;

In [52]:
query = '''SELECT b.title,
            COUNT(re.review_id) as review_number,
            ROUND(AVG(ra.rating),1) as rating_mean
FROM books as b
LEFT JOIN ratings AS ra ON ra.book_id=b.book_id
LEFT JOIN reviews AS re ON re.book_id=b.book_id
GROUP BY b.book_id
ORDER BY  review_number DESC, rating_mean DESC
'''
con=engine.connect()
pd.io.sql.read_sql(sql=text(query), con = con)

title  review_number  \
0                              Twilight (Twilight  #1)           1120   
1                  The Hobbit  or There and Back Again            528   
2                               The Catcher in the Rye            516   
3    Harry Potter and the Prisoner of Azkaban (Harr...            492   
4    Harry Potter and the Chamber of Secrets (Harry...            480   
..                                                 ...            ...   
995                               Leonardo's Notebooks              0   
996  Disney's Beauty and the Beast (A Little Golden...              0   
997                          Essential Tales and Poems              0   
998    Anne Rice's The Vampire Lestat: A Graphic Novel              0   
999                            The Natural Way to Draw              0   

     rating_mean  
0            3.7  
1            4.1  
2            3.8  
3            4.4  
4            4.3  
..           ...  
995          4.0  
996          4.0  
997          4.0  
998          3.7  
999          3.0  

[1000 rows x 3 columns]

В данных 1000 книг. Оценки колеблятся от 5.0 до 1.5. Число обзоров - от 1120 до 0.

### Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;

In [53]:
query = ''' SELECT p.publisher,
            COUNT(b.num_pages) as big_books
            
FROM books as b
INNER JOIN publishers AS p ON p.publisher_id=b.publisher_id
WHERE b.num_pages>50
GROUP BY p.publisher
ORDER BY big_books DESC
LIMIT 1
'''
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'Самое производительное издательство - {result.values[0][0]}.\nОно выпустило {result.values[0][1]} книги толщиной более 50 страниц.')

Самое производительное издательство - Penguin Books.
Оно выпустило 42 книги толщиной более 50 страниц.


### Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;

In [54]:
query = ''' 
SELECT author,
    AVG(r.rating) AS avg_rating

FROM books AS b
    INNER JOIN authors AS a on a.author_id=b.author_id
    INNER JOIN ratings AS r on r.book_id=b.book_id

WHERE b.book_id IN(
    SELECT  distinct book_id
    FROM ratings
        GROUP BY book_id
        HAVING COUNT(rating)>50)

    GROUP BY author
    ORDER BY avg_rating DESC
    LIMIT 1
''' 
con=engine.connect()
result=pd.io.sql.read_sql(sql=text(query), con = con)
print(f'Автор {result.values[0][0]} - любимец публики.\nСредний рейтинг этого автора по всем книгам - {result.values[0][1]}.\nЭто самое большое значение среди всех авторов книг, которым поставили 50 и более оценок.')

Автор J.K. Rowling/Mary GrandPré - любимец публики.
Средний рейтинг этого автора по всем книгам - 4.287096774193548.
Это самое большое значение среди всех авторов книг, которым поставили 50 и более оценок.


### Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [55]:
query = '''
SELECT username,
    AVG(counted_reviews) AS avg_reviews
FROM (
    SELECT COUNT(review_id) as counted_reviews,
           username
    FROM reviews
        GROUP BY username
) as re
WHERE username IN (
SELECT  username        
    FROM ratings
        GROUP BY username
        HAVING COUNT(username)>48
)
     
GROUP BY username  
ORDER BY avg_reviews DESC
        
        ''' 
con=engine.connect()
print('Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок:')
pd.io.sql.read_sql(sql=text(query), con = con)   

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок:


username  avg_reviews
0          susan85         29.0
1      sfitzgerald         28.0
2       martinadam         27.0
3        richard89         26.0
4      lesliegibbs         26.0
5   jennifermiller         25.0
6   shermannatalie         24.0
7         walter49         23.0
8      andreaeaton         22.0
9           paul88         22.0
10  vanessagardner         22.0
11         vcarter         20.0
12          xdavis         18.0

## Вывод:

В базе содержится 819 книг, вышедших после 1.01.2000.

В данных 1000 книг. Оценки колеблятся от 5.0 до 1.5. Число обзоров - от 1120 до 0.

Самое производительное издательство - Penguin Books.
Оно выпустило 42 книги толщиной более 50 страниц.

Автор J.K. Rowling/Mary GrandPré - любимец публики.
Средний рейтинг этого автора по всем книгам - 4.23.
Это самое большое значение среди всех авторов книг, которым поставили 50 и более оценок.

Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок колеблется от 18 до 29